ankaios_sdk/ankaios.py

```diff
@@ -0,0 +1,662 @@
+        DEBUG (int): Debug log level.
+    """
+    FATAL = logging.FATAL
+    ERROR = logging.ERROR
+    WARN = logging.WARN
+    INFO = logging.INFO
+    DEBUG = logging.DEBUG
+
+
+# pylint: disable=too-many-public-methods
+class Ankaios:
+    """
+    A class to interact with the Ankaios control interface. It provides
+    the functionality to interact with the Ankaios control interface
+    by sending requests.
+
+    Attributes:
+        ANKAIOS_CONTROL_INTERFACE_BASE_PATH (str): The base path for the
+            Ankaios control interface.
+        DEFAULT_TIMEOUT (int): The default timeout, if not manually provided.
+        logger (logging.Logger): The logger for the Ankaios class.
+        path (str): The path to the control interface.
+    """
+    ANKAIOS_CONTROL_INTERFACE_BASE_PATH = "/run/ankaios/control_interface"
+    DEFAULT_TIMEOUT = 5.0
+
+    def __init__(self) -> None:
+        """Initialize the Ankaios object."""
+        self.logger = None
+        self.path = self.ANKAIOS_CONTROL_INTERFACE_BASE_PATH
+
+        self._read_thread = None
+        self._connected = False
+        self._responses_lock = threading.Lock()
+        self._responses: dict[str, ResponseEvent] = {}
+
+        self._create_logger()
+
+    def __enter__(self) -> "Ankaios":
+        """
+        Connect to the control interface.
+
+        Returns:
+            Ankaios: The Ankaios object.
+        """
+        self.connect()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback) -> None:
+        """
+        Disconnect from the control interface.
+
+        Args:
+            exc_type (type): The exception type.
+            exc_value (Exception): The exception instance.
+            traceback (traceback): The traceback object.
+        """
+        if exc_type is not None:  # pragma: no cover
+            self.logger.error("An exception occurred: %s, %s, %s",
+                              exc_type, exc_value, traceback)
+        self.disconnect()
+
+    def _create_logger(self) -> None:
+        """Create a logger with custom format and default log level."""
+        formatter = logging.Formatter('%(asctime)s %(message)s',
+                                      datefmt="%FT%TZ")
+        self.logger = logging.getLogger("Ankaios logger")
+        handler = logging.StreamHandler()
+        handler.setFormatter(formatter)
+        self.logger.addHandler(handler)
+        self.set_logger_level(AnkaiosLogLevel.INFO)
+
+    def _read_from_control_interface(self) -> None:
+        """
+        Reads from the control interface input fifo and saves the response.
+        This is meant to be run in a separate thread.
+        It reads the response from the control interface and saves it in the
+        responses dictionary, by triggering the corresponding ResponseEvent.
+        """
+        # pylint: disable=consider-using-with
+        f = open(f"{self.ANKAIOS_CONTROL_INTERFACE_BASE_PATH}/input", "rb")
```

Comment:
What do you have in mind regarding the input_fifo name?
The only reason for not using `with` is that there would be too many indentations in the code.

Reply:
The file that we open is called just "f" and is actually the input_fifo. I know the naming as "f" is done often in python, but would be nice to name it properly here as the implementation will also act as how to for somebody that does not want to use the SDK.

As for the "with", if we cannot open the file, it would look ugly. You can try it out with a workload not having Control Interface access 😉

Reply:
Moved the open in the `try except` so that no error might slip and also renamed `f` into `input_fifo`.